llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp

```diff
@@ -500,6 +500,45 @@ static bool isSelect01(const APInt &C1I, const APInt &C2I) {
   return C1I.isOne() || C1I.isAllOnes() || C2I.isOne() || C2I.isAllOnes();
 }
 
+/// Try to simplify a select instruction when the user of its select user
+/// indicates the condition.
+static bool simplifySeqSelectWithSameCond(SelectInst &SI,
+                                          const SimplifyQuery &SQ,
+                                          InstCombinerImpl &IC) {
+  Value *CondVal = SI.getCondition();
+  if (isa<Constant>(CondVal))
+    return false;
+
+  auto TrysimplifySeqSelect = [=, &SI, &IC](unsigned OpIndex) {
+    assert((OpIndex == 1 || OpIndex == 2) && "Unexpected operand index");
+    Type *CondType = CondVal->getType();
+    SelectInst *SINext = &SI;
+    Value *ValOp = SINext->getOperand(OpIndex);
+    Value *CondNext;
+    while (match(ValOp, m_Select(m_Value(CondNext), m_Value(), m_Value()))) {
+      if (CondNext == CondVal) {
+        IC.replaceOperand(*SINext, OpIndex,
+                          cast<SelectInst>(ValOp)->getOperand(OpIndex));
+        return true;
+      }
+
+      SINext = cast<SelectInst>(ValOp);
+      ValOp = SINext->getOperand(OpIndex);
+    }
```

Comment:
Im not sure concerned about this, I think it may be slight better to continue looping until no more select and have a changed boolean.
Take a case like:
```
define i8 @foo(i1 %cond0, i1 %cond1, i8 %w, i8 %x, i8 %y, i8 %z, i8 %a) {
  %sel0 = select i1 %cond0, i8 %w, i8 %x
  %sel1 = select i1 %cond1, i8 %y, i8 %sel0
  %sel2 = select i1 %cond0, i8 %sel1, i8 %z
  %sel3 = select i1 %cond1, i8 %sel2, i8 %a
  ret i8 %sel3
}
```

I think what happens here is you replace `%sel2 = select i1 %cond0, i8 %sel1, i8 %z` -> `%sel2 = select i1 %cond0, i8 %y, i8 %z`. But after that you won't be able to replace ` %sel1 = select i1 %cond1, i8 %y, i8 %sel0` -> ` %sel1 = select i1 %cond1, i8 %y, i8 %x` b.c you won't match any `select` arm in `%sel2`.

I'm not really sure if this ever comes up, I don't think its a must fix, just a thought.

Reply:
Yes, I think it need anther patch to handle **match(CondNext, m_Not(CondVal))** and flip OpIndex for the replacement as your previous commit, https://alive2.llvm.org/ce/z/yGCRqB.

Reply:
Thansk, need a seperate patch to improve the case  https://alive2.llvm.org/ce/z/YGZPFa